src/lib/statistics/chunk_statistics/counting_quotient_filter.hpp

```diff
@@ -0,0 +1,58 @@
+#pragma once
+
+#include <vector>
+#include <string>
+
+#include "cqf2.hpp"
+#include "cqf4.hpp"
+#include "cqf8.hpp"
+#include "cqf16.hpp"
+#include "cqf32.hpp"
+#include "types.hpp"
+
+#include "storage/base_column.hpp"
+#include "abstract_filter.hpp"
+
+
+
+
+namespace opossum {
+
+/* Counting Quotient Filters allow you to keep track of which values are present in a column and how often. Filters
+work approximately. If a membership query yields a positive result, the value is probably present but there is
+a chance of a false positive. If the query delivers a negative result, the item is guaranteed to not be contained.
+In the same way, items can be over counted but not under counted.
+CQF can be configured with quotient size, which determines the number of slots, and the remainder size, which
+corresponds to the slot size. At this time, the remainder size must be 2, 4, 8, 16 or 32. */
+
+template <typename ElementType>
+class CountingQuotientFilter : AbstractFilter {
+ public:
+  CountingQuotientFilter(uint8_t quotient_bits, uint8_t remainder_bits);
+  virtual ~CountingQuotientFilter();
+  void insert(ElementType value, uint64_t count);
+  void insert(ElementType value);
+  void populate(std::shared_ptr<const BaseColumn> column);
+  uint64_t count(ElementType value) const;
+  uint64_t count_all_type(AllTypeVariant value) const;
+  uint64_t memory_consumption() const;
+  double load_factor() const;
+  bool is_full() const;
+
+  bool can_prune(const AllTypeVariant& value, const PredicateCondition predicate_type) const override;
+
+ private:
+  std::optional<gqf2::quotient_filter> _quotient_filter2;

```

Comment:
Just from the looks of it, using `boost::variant` and `boost::apply_visitor` for these would save you a lot of error-prone if-elseif-else blocks

Reply:
@mrks : do you know if we can safely use `std::optional` here? I don't remember the reason to not use it in the hash join.